### ISSUE TITLE
useDefaultValueForNull not respected during compaction

### ISSUE THREAD

**Issue by WilliamWhispell** [apache/druid#8221]
Reproducer:

1. set druid.generic.useDefaultValueForNull=false in common.runtime.properties and start druid (micro-quick-start)
2. Create a data source with double dimensions and load data for two different days with appendToExisting set to true - this creates at least two segments. Load some empty values so that you have null values.
3. Query the datasource and ensure you get back null for some dimensions
4. Run a compact task on the data source

After the compaction all of the null values are converted to the default values. I checked the peon logs for the compact process and found:

2019-08-01T17:57:32,062 INFO [main] org.apache.druid.cli.CliPeon - * druid.extensions.loadList: ["druid-kafka-indexing-service", "druid-datasketches", "postgresql-metadata-storage", "druid-s3-extensions"]
2019-08-01T17:57:32,062 INFO [main] org.apache.druid.cli.CliPeon - * druid.generic.useDefaultValueForNull: false
2019-08-01T17:57:32,062 INFO [main] org.apache.druid.cli.CliPeon - * druid.host: localhost

Do I need to set the useDefaultValueForNull somewhere else too?

@nishantmonu51

**Comment by WilliamWhispell** [comment]
I'm using s3 as a persistent store for the segments.. not sure if that has anything to do with it.

**Comment by WilliamWhispell** [comment]
Today I wanted to see if this was from using s3, so I loaded the data with nulls, then I manually deleted the local segment-cache. I restarted druid and waited for it two fetch the two segments from s3. Once they were available I found that nulls were still returned as expected.

I'll be attempting to debug the compaction process next.

**Comment by WilliamWhispell** [comment]
Here we write the total size of both the vector and the null bitmap.

https://github.com/apache/incubator-druid/blob/3ae563263a23000560749071d262727d47296856/processing/src/main/java/org/apache/druid/segment/DoubleColumnSerializerV2.java#L130

And here we move out byte iterator that number of bytes, then we look for the null bitmap, but obviously we passed it.

https://github.com/apache/incubator-druid/blob/e1a745717e0b008ca76a0a27e96d3097fc48a8f9/processing/src/main/java/org/apache/druid/segment/serde/DoubleNumericColumnPartSerdeV2.java#L153

**Comment by WilliamWhispell** [comment]
Okay, so I've fixed this locally, but not familiar enough with the code yet to open a PR.

During compaction, we use the ingestsegmentfirehose, and it calls lookupName on the dimensionselector. 

https://github.com/apache/incubator-druid/blob/master/server/src/main/java/org/apache/druid/segment/realtime/firehose/IngestSegmentFirehose.java#L146

I noticed we were getting back 0.0 here, which is the value serialized, but it does not look at the bitmap index. I found that the doublewrappingdimensionselector does not check IsNull on the selector, but directly calls getDouble, this violates the asset in the ColumnarDoubles getDouble function. 

https://github.com/apache/incubator-druid/blob/3ae563263a23000560749071d262727d47296856/processing/src/main/java/org/apache/druid/segment/DoubleWrappingDimensionSelector.java#L44

So to fix locally, I changed:

@Nullable
  @Override
  protected String getValue()
  {
    if (extractionFn == null) {
      return String.valueOf(selector.getDouble());
    } else {
      return extractionFn.apply(selector.getDouble());
    }
  }

to first check selector.isNull() and to return null if it is, instead of calling getDouble. After this change, nulls remain null after compacting segments for doubles, I'll need to check the other types to see if they're affected too.

@nishantmonu51
